lib/type_checker/infer.ml

```diff
@@ -302,39 +459,66 @@ module Structure = struct
     let_ cvar #= (poly_scheme (quantifiers @. tt @=> type_)) ~in_:c
   ;;
 
-  let infer_type_decl ~env { type_decl_name; type_decl_params; type_decl_kind } =
-    let type_kind =
+  let infer_type_decl ~(env : Env.t) { type_decl_name; type_decl_params; type_decl_kind } =
+    let open Or_error.Let_syntax in
+    (* Create a fresh identifier for the type *)
+    let type_ident =
+      Adt.Type_ident.create ~id_source:env.id_source ~name:(type_decl_name :> string) ()
+    in
+    (* Convert the declaration kind *)
+    let%map type_kind =
       match type_decl_kind with
-      | Type_decl_abstract -> Adt.Type_abstract
+      | Type_decl_abstract -> return Adt.Type_abstract
       | Type_decl_variant constr_decls ->
-        let constructor_type =
-          Type_constr
-            ( List.map type_decl_params ~f:(fun type_var -> Type_var type_var)
-            , type_decl_name )
-        in
-        let constr_decls =
-          List.map constr_decls ~f:(fun { constructor_name; constructor_arg } ->
-            { Adt.constructor_name
-            ; constructor_alphas = type_decl_params
-            ; constructor_type
-            ; constructor_arg
-            })
-        in
-        Type_variant constr_decls
+        Env.declare_type env ~type_name:type_decl_name ~type_ident ~in_:(fun env ->
```

Comment:
This isn't entirely correct for mutually recursive types -- we need to declare the entire block of types, collect all the declarations to add and the update the env. Additionally it would be nice to check the arity of types when doing inference here

Reply:
Will fix this in a follow up PR

Reply:
See #16